app/notifiers/notifier/base.rb

```diff
@@ -0,0 +1,84 @@
+# The base class of notifiers. This is meant to be called by the Notifications Framework
+#
+# @api notifications
+class Notifier::Base
+  include ApplicationNotificationsConcern
+
+  class ActivityWrapper < SimpleDelegator
+    # Send notifications according to input type and recipient
+    #
+    # @param [Object] recipient The recipient of the notification
+    # @param [Symbol] type The type of notification
+    def notify(recipient, type)
+      super(recipient, type).tap do |notification|
+        case recipient
+        when Course
+          @pending_emails += @notifier.send(:email_course, notification) if type == :email
+        when User
+          @pending_emails << @notifier.send(:email_user, notification) if type == :email
+        end
+      end
+
+      self
+    end
+
+    # Save activity and send out pending emails
+    def save(*)
+      super.tap do |result|
+        @pending_emails.each(&:deliver_later) if result
+      end
+    end
+
+    private
+
+    def initialize(notifier, activity) #:nodoc:
+      super(activity)
+      @notifier = notifier
+      @pending_emails = []
+    end
+  end
+
+  class << self
+    # Call back method which is used to call method directly though the notifier class. This is
```

Comment:
This is not a callback for client code, which is what your comment seems to imply.

This is to allow client code to call create notifications without explicitly creating notifiers.